float64 x
float64 y
bool reset
bool firstattach
bool pred
